### PR TITLE
Sorted imports 2

### PR DESCRIPTION
**Pull request by expipiplus1**
I've merged the ghc 8.10 branch against the sorted-import branch and added some basic tests (Had to add a blank line function to the test suite)

I'm not sure if I'm being dense, but are there any tests where one enters the initial and expected source files, or is every test a "passthrough" test. i.e. would the identity function pass the testsuite?!

(I've targeted this against the ghc-8.10 branch because it includes those commits too)

The only required checklist items on @lspitzner's PR were:

- tests (I'm not sure if these ones here are sufficient)
- testing on a large codebase (I've been running this daily for ages without any issues whatsoever (including on some very large codebases))

(reopened from https://github.com/lspitzner/brittany/pull/325)

### CODE REVIEW

File: src-literatetests/30-tests-context-free.blt [hunk 1]
```diff
@@ -796,22 +796,22 @@ import Prelude as X
     )
 
 #test long-module-name-simple
-import TestJustShortEnoughModuleNameLikeThisOne ()
-import TestJustAbitToLongModuleNameLikeThisOneIs ()
 import MoreThanSufficientlyLongModuleNameWithSome
-  (items, that, will, not, fit, inA, compact, layout)
+  (compact, fit, inA, items, layout, not, that, will)
```
Comment (tfausak) [inline src-literatetests/30-tests-context-free.blt:800]: 😆 These are funny now that they're sorted.

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs [hunk 2]
```diff
@@ -99,17 +103,83 @@ layoutIE lie@(L _ ie) = docWrapNode lie $ case ie of
...
+  mergeGroup items@[_] = items
+  mergeGroup items     = if
+    | all isProperIEThing items -> [List.foldl1' thingFolder items]
+    | all isIEVar items         -> [List.foldl1' thingFolder items]
+    | otherwise                 -> items
+  -- proper means that if it is a ThingWith, it does not contain a wildcard
+  -- (because I don't know what a wildcard means if it is not already a
+  -- IEThingAll).
+  isProperIEThing :: LIE GhcPs -> Bool
+#if MIN_VERSION_ghc(8,6,0) /* 8.6+ */
```
Comment (tfausak) [inline src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:137]: We can drop this CPP now.

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs [hunk 3]
```diff
@@ -99,17 +103,83 @@ layoutIE lie@(L _ ie) = docWrapNode lie $ case ie of
...
+  isIEVar = \case
+    L _ IEVar{} -> True
+    _           -> False
+  thingFolder :: LIE GhcPs -> LIE GhcPs -> LIE GhcPs
+  thingFolder l1@(L _ IEVar{}     ) _                     = l1
+  thingFolder l1@(L _ IEThingAll{}) _                     = l1
+  thingFolder _                     l2@(L _ IEThingAll{}) = l2
+  thingFolder l1                    (   L _ IEThingAbs{}) = l1
+  thingFolder (L _ IEThingAbs{})    l2                    = l2
+#if MIN_VERSION_ghc(8,6,0) /* 8.6+ */
```
Comment (tfausak) [inline src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:160]: Same with this.

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs [hunk 4]
```diff
@@ -124,26 +194,73 @@ layoutAnnAndSepLLIEs llies@(L _ lies) = do
...
+      addAlternative
+        $  docPar (docSetBaseY $ docSeq [docParenLSep, ieDsH])
+        $  docLines
+        $  ieDsT
+        ++ [docParenR]
+
+-- | Returns a "fingerprint string", not a full text representation, nor even
+-- a source code representation of this syntax node.
+-- Used for sorting, not for printing the formatter's output source code.
+#if MIN_VERSION_ghc(8,2,0)
```
Comment (tfausak) [inline src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:222]: And this.

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs [hunk 5]
```diff
@@ -124,26 +194,73 @@ layoutAnnAndSepLLIEs llies@(L _ lies) = do
...
+  L _ (IEType    n) -> lrdrNameToText n
+#else
+wrappedNameToText :: Located RdrName -> Text
+wrappedNameToText = lrdrNameToText
+#endif
+-- | Returns a "fingerprint string", not a full text representation, nor even
+-- a source code representation of this syntax node.
+-- Used for sorting, not for printing the formatter's output source code.
+lieToText :: LIE GhcPs -> Text
+#if MIN_VERSION_ghc(8,6,0) /* 8.6+ */
```
Comment (tfausak) [inline src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:236]: And this.

File: src-literatetests/30-tests-context-free.blt (hunk 1 above)
Comment (expipiplus1) [reply src-literatetests/30-tests-context-free.blt:800]: If you can think of an alphabetical sentence with the same meaning we should use that!

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs (hunk 5 above)
Comment (expipiplus1) [reply src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:236]: done

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs (hunk 4 above)
Comment (expipiplus1) [reply src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:222]: done

File: src/Language/Haskell/Brittany/Internal/Layouters/IE.hs (hunk 3 above)
Comment (expipiplus1) [reply src/Language/Haskell/Brittany/Internal/Layouters/IE.hs:160]: done

**Comment by tfausak** [comment]
> are there any tests where one enters the initial and expected source files

Not that I know of, but it seems like there should be!

**Review by tfausak — APPROVED** [review]
I can't really speak to the code, but the result (sorting imports) looks great! Thanks for sticking with this and getting it over the line.

**Comment by expipiplus1** [comment]
Thanks!

**Pull request merged**